### DISCUSSION TITLE
It is suggested to improve usability of dev::ServiceRequest.

### DISCUSSION THREAD

**Discussion by myyrakle** [actix/actix-web#1863] [Ideas] [answered]
When writing middleware, it is common to disassemble and use ServiceRequest objects.
However, there is a problem when recreating the ServiceRequest object with the disassembled HttpRequest and Payload.
![image](https://user-images.githubusercontent.com/16988115/103331244-4fe06800-4aa8-11eb-880b-6a94d145e572.png)
![image](https://user-images.githubusercontent.com/16988115/103331262-65559200-4aa8-11eb-895d-7e7d3e94698e.png)
To unwrap, Result's Error type must implement Debug trait, but it cannot.
This is because (HttpRequest, Payload) is returned as the error type.

The best way to solve this now is to use unwrap by converting Result to Option.
![image](https://user-images.githubusercontent.com/16988115/103331305-9d5cd500-4aa8-11eb-9a9e-4563da441159.png)

However, it is not intuitive and inconvenient.
How about changing the type by defining a dedicated error type, or changing the result to an option altogether?

**Comment by fakeshadow** [comment, marked as answer]
It returns the ownership back to you so you don't lose them forever.

In general don't use unwrap for error handling. You want to handle the error properly.

**Reply by myyrakle** [reply]
However, if you want to use unwrap_or, etc., it is difficult to create a default value.
If so, what method do you recommend to use when fetching values?

**Reply by fakeshadow** [reply]
let a = Err(());

if a.is_err() {
    // do something
}

**Reply by fakeshadow** [reply]
In fact this is what unwrap_or is
```rust
    pub fn unwrap_or(self, default: T) -> T {
        match self {
            Ok(t) => t,
            Err(_) => default,
        }
    }
```

**Reply by myyrakle** [reply]
![image](https://user-images.githubusercontent.com/16988115/103332026-dba7c380-4aab-11eb-8f3a-677c779a70a4.png)
Even with error handling as per your advice, I will need to unwrap to get the value eventually.
What am I doing wrong?

**Reply by fakeshadow** [reply]
You are not handling error. You literly just try to unwrap it again.
Handling error means return an error according to the Service::Error type or construct a ServiceResponse which would represent the error you want to return to client.

**Reply by fakeshadow** [reply]
```rust
    fn handle_error(req: HttpRequest) -> Result<ServiceResponse, actix_web::Error> {
        match ServiceRequest::from_parts(req, Payload::None) {
            Ok(req) => {
                Ok(req.into_response("OK"))
            },
            Err((req, _)) => {
                Ok(ServiceResponse::from_err(actix_web::error::ErrorInternalServerError("Error"), req))
            }
        }
    }
```

**Reply by myyrakle** [reply]
Oh... thank you. I was thinking wrong.
